remotefs: Added sync to writes and OpenFile

gupta-ak opened the pull request:
Added a `sync` for writes, so that writes from `docker cp`  will always persistent to the container.

Also added a `OpenFile` command to support cases where docker calls `Open/OpenFile`. Since the launch process architecture makes it hard for `remotefs` to keep state, the `OpenFile` command is mainly a permission check. After calling `OpenFile`, docker can call `ReadFile` to get the file on the host, use it as a normal `*os.File` and then do `WriteFile` to send the file to the VM to write to the VHDX.

msftclas commented:
@gupta-ak,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_

jstarks commented on service/gcsutils/remotefs/utils.go:
```diff
@@ -146,3 +147,9 @@ func WriteTarOptions(w io.Writer, opts *archive.TarOptions) error {
+// Sync syncs the filesystem so that all writes buffered in the kernel are flushed
+// to disk.
+func Sync() error {
+	return exec.Command("sync").Run()
```
This is a pretty heavy hammer. Can't we implement this just to fsync the files that have been modified?

gupta-ak replied:
Yeah, I can try with `fsync`

jstarks commented on service/gcsutils/remotefs/remotefs_os.go:
```diff
@@ -239,7 +258,45 @@ func Mkfifo(in io.Reader, out io.Writer, args []string) error {
+}
+
+// OpenFile works like os.OpenFile. Since the GCS process calling structure
+// does not enable us to keep state, OpenFile doesn't return a handle to the file.
```
:(. What do we need to do to improve this?

gupta-ak replied:
I'm not sure. I guess a workaround can be for docker to call `remotefs OpenFile` and then have a small protocol for reading/writing/seeking/closing through `stdin` & `stdout`

gupta-ak commented:
@jstarks Updated

1) Removed the `Sync()` stuff because it's not necessary. `umount` already does a `sync`, so the data should be flushed if `docker` calls `graphdriver.Put()`. The reason why I wasn't seeing it being synced because I forgot to wait for `umount` to complete.
2) I changed the OpenFile implementation to have a "mini fileserver" protocol in to support `*os.File Read/Write/Close`, so the client now keeps a handle to the `remotefs` process and issues a `Close command` when they're done with the file.
